File: TasteKit/tastekit/packages/core/drift/consolidator.ts
/**
 * Memory Consolidator
 * 
 * Consolidates memory based on salience and retention policy.
 * 
 * Note: This is a policy-level consolidator. Actual memory consolidation
 * would be performed by the runtime adapter.
 */

export interface ConsolidationPlan {
  timestamp: string;
  memories_to_keep: string[];
  memories_to_prune: string[];
  memories_to_merge: Array<{
    source_ids: string[];
    merged_content: string;
  }>;
}

export class MemoryConsolidator {
  /**
   * Generate a consolidation plan based on salience scores and retention policy.
   * 
   * This is a stub implementation. A full implementation would:
   * 1. Fetch memories from runtime adapter
   * 2. Calculate salience scores
   * 3. Apply retention policy
   * 4. Identify merge candidates
   * 5. Generate plan for runtime to execute
   */
  generateConsolidationPlan(
    memories: Array<{ id: string; content: string; salience: number; timestamp: string }>,
    retentionDays: number
  ): ConsolidationPlan {
    const now = new Date();
    const cutoffDate = new Date(now.getTime() - retentionDays * 24 * 60 * 60 * 1000);
    
    const toKeep: string[] = [];
    const toPrune: string[] = [];
    
    for (const memory of memories) {
      const memoryDate = new Date(memory.timestamp);
      
      // Keep if recent or high salience
      if (memoryDate >= cutoffDate || memory.salience > 0.7) {
        toKeep.push(memory.id);
      } else {
        toPrune.push(memory.id);
      }
    }
    
    return {
      timestamp: now.toISOString(),
      memories_to_keep: toKeep,
      memories_to_prune: toPrune,
      memories_to_merge: [], // TODO: Implement merge detection
    };
  }
}

File: AutoClaw/internal/drift/consolidator.go
package drift

import (
	"strings"
	"time"
)

// MemoryEntry represents a single memory for consolidation.
type MemoryEntry struct {
	ID        string  `json:"id"`
	Content   string  `json:"content"`
	Salience  float64 `json:"salience"`  // 0.0-1.0
	Timestamp string  `json:"timestamp"` // ISO8601
}

// MergeGroup describes memories to be consolidated.
type MergeGroup struct {
	SourceIDs     []string `json:"source_ids"`
	MergedContent string   `json:"merged_content"`
}

// ConsolidationPlan describes what to keep, prune, and merge.
type ConsolidationPlan struct {
	Timestamp      string       `json:"timestamp"`
	MemoriesToKeep []string     `json:"memories_to_keep"`
	MemoriesToPrune []string    `json:"memories_to_prune"`
	MemoriesToMerge []MergeGroup `json:"memories_to_merge"`
}

// Consolidator manages memory lifecycle.
// Algorithm matches TasteKit:
//   - Keep if recent OR salience > 0.7
//   - Merge if Jaccard similarity >= 0.6
//   - Highest salience = primary in merge
type Consolidator struct{}

// NewConsolidator creates a memory consolidator.
func NewConsolidator() *Consolidator {
	return &Consolidator{}
}

// GenerateConsolidationPlan analyzes memories and returns a plan.
func (c *Consolidator) GenerateConsolidationPlan(memories []MemoryEntry, retentionDays int) ConsolidationPlan {
	now := time.Now().UTC()
	cutoff := now.AddDate(0, 0, -retentionDays)

	plan := ConsolidationPlan{
		Timestamp: now.Format(time.RFC3339),
	}

	// Step 1: Retention filter.
	var toKeep []MemoryEntry
	for _, m := range memories {
		t, err := time.Parse(time.RFC3339, m.Timestamp)
		if err != nil {
			// If we can't parse the timestamp, keep it to be safe.
			toKeep = append(toKeep, m)
			continue
		}
		if !t.Before(cutoff) || m.Salience > 0.7 {
			toKeep = append(toKeep, m)
		} else {
			plan.MemoriesToPrune = append(plan.MemoriesToPrune, m.ID)
		}
	}

	// Step 2: Similarity-based merge detection (Jaccard >= 0.6).
	merged := make(map[string]bool)
	for i := 0; i < len(toKeep); i++ {
		if merged[toKeep[i].ID] {
			continue
		}
		group := []MemoryEntry{toKeep[i]}
		for j := i + 1; j < len(toKeep); j++ {
			if merged[toKeep[j].ID] {
				continue
			}
			if JaccardSimilarity(toKeep[i].Content, toKeep[j].Content) >= 0.6 {
				group = append(group, toKeep[j])
				merged[toKeep[j].ID] = true
			}
		}

		if len(group) > 1 {
			merged[toKeep[i].ID] = true
			mg := c.mergeGroup(group)
			plan.MemoriesToMerge = append(plan.MemoriesToMerge, mg)
		}
	}

	// Step 3: Collect standalone (unmerged) memories.
	for _, m := range toKeep {
		if !merged[m.ID] {
			plan.MemoriesToKeep = append(plan.MemoriesToKeep, m.ID)
		}
	}

	return plan
}

// mergeGroup consolidates similar memories. Highest salience becomes primary;
// unique content from others is appended.
func (c *Consolidator) mergeGroup(group []MemoryEntry) MergeGroup {
	// Sort by salience descending (simple selection of max).
	maxIdx := 0
	for i, m := range group {
		if m.Salience > group[maxIdx].Salience {
			maxIdx = i
		}
	}
	// Move primary to front.
	group[0], group[maxIdx] = group[maxIdx], group[0]

	primary := group[0]
	var ids []string
	for _, m := range group {
		ids = append(ids, m.ID)
	}

	var additionalDetails []string
	for _, m := range group[1:] {
		unique := ExtractUniqueContent(m.Content, primary.Content)
		if unique != "" {
			additionalDetails = append(additionalDetails, unique)
		}
	}

	content := primary.Content
	if len(additionalDetails) > 0 {
		content += "\n\nAdditional context: " + strings.Join(additionalDetails, "; ")
	}

	return MergeGroup{
		SourceIDs:     ids,
		MergedContent: content,
	}
}
